pallets/ds-maps/src/lib.rs

```diff
@@ -0,0 +1,192 @@
+}
+
+decl_storage!{
+    // A unique name is used to ensure that the pallet's storage items are isolated.
+    // This name may be updated, but each pallet in the runtime must use a unique name.
+    // ---------------------------------vvvvvvvvvvvv
+    trait Store for Module<T: Trait> as DSMapsModule {
+        // MAX is 4_294_967_295. Change if required more.
+        TotalBoxes get(fn total_boxes): u32;    
+
+        CityMap get(fn map_data): 
+            map hasher(blake2_128_concat) u32 => ZoneOf<T>;
+    }
+}
+pub type ZoneOf<T> = Zone<<T as Trait>::Point>;
+
+// Pallets use events to inform users when important changes are made.
+// https://substrate.dev/docs/en/knowledgebase/runtime/events
+decl_event!(
+    pub enum Event<T>
+    where
+        AccountId = <T as frame_system::Trait>::AccountId,
+        Coord = <T as Trait>::Coord,
+    {
+        // Event documentation should end with an array that provides descriptive names for event parameters.
+        /// TODO add more meta/remove
+        MapInitialized(Coord),
+        /// New account has been created [zone number, its type], TODO later add printing coords
+        ZoneCreated(u32, AccountId, ZoneType),
+    }
+);
+
+// Errors inform users that something went wrong.
+// learn more https://substrate.dev/docs/en/knowledgebase/runtime/errors
+decl_error! {
+    pub enum Error for Module<T: Trait> {
+        /// Error names should be descriptive.
+        NoneValue,
+        /// Operation is not valid
+        InvalidAction,
+        /// Incorrect data provided
+        InvalidData,
+        /// Origin do not have sufficient privileges to perform the operation
+        NotAuthorized,
+        /// Account doesn't exist
+        NotExists,
+        // add additional errors below
+    }
+}
+
+// Dispatchable functions allows users to interact with the pallet and invoke state changes.
+// These functions materialize as "extrinsics", which are often compared to transactions.
+// Dispatchable functions must be annotated with a weight and must return a DispatchResult.
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        // Errors must be initialized if they are used by the pallet.
+        type Error = Error<T>;
+
+        // Events must be initialized if they are used by the pallet.
+        fn deposit_event() = default;
+
+        #[weight = <T as Trait>::WeightInfo::zone_add()]
+        pub fn zone_add(origin, 
+                        zone_type: ZoneType, 
+                        bounding_box: Box3D<T::Point>) -> dispatch::DispatchResult {
+            let who = ensure_signed(origin)?;
+            // TODO implement inverted index, so we will not store same zones twice
+            ensure!(<accounts::Module<T>>::account_is(&who, REGISTRAR_ROLE.into()), Error::<T>::NotAuthorized);
+            
+            let id = <TotalBoxes>::get();
+            let zone = ZoneOf::<T>::new(id, zone_type.clone(), bounding_box);
```

Comment:
ZoneType is a light-weight enum, without extra data associated with its variants, so it can have `Copy` trait so that we don't need to use `.clone()` everywhere.